### PR TITLE
Api broker

### PR DESCRIPTION
**Pull request by badaniya**
Pyswitchlib API Daemon (API Broker)
 
Prerequistes:
========
- Must have root privilege to have pyswitchlib-api-daemon service installed and for daemon restartability.

Comments:
=======
1. The pyswitchlib-api-daemon service is installed and registered during pip install of the pyswitchlib package (Currently pip installed from github source).
2. The pyswitchlib-api-daemon service can start one or more pyswitchlib api daemons depending on the service configuration file.
3. The pyswitchlib daemon(s) will be started or restarted when a pip install is done.
4. A default API daemon will be started if the config file does not match any configured prefixes.  For example if the pyswitchlib-api-daemon service is installed on to the host's base python.   

Pyswitchlib-api-daemon Service Configuration:
==============================
The pyswitchlib-api-daemon service is installed with a default configuration file.  If a working configuration file does not exist, then the default configuration file will be copied over and setup as the initial working configuration file.  If a working configuration file already exits during an upgrade/install, then the working configuration file is preserved and will be used to start/restart the api daemon(s).

    - Default config file is installed at /etc/pyswitchlib/pyswitchlib.conf.default

    - Working config file is at /etc/pyswitchlib/pyswitchlib.conf.  If a working config file does not exist then the default config file will be copied and used as the working.

    - The config file specifies how many api daemons will be launched and which sys.prefixes should be associated to the daemon.  When multiple sys.prefixes or virtualenv prefixes are associated to the api daemon (delimited by a colon), then any pyswitchlib assets created under these virtualenvs will utilize this single api daemon.

    -  The default config specifies two api daemons:
        1.  api_daemon_bwc_topology:  The associated sys.prefix is the bwc-topology virtualenv path
        2.  api_deamon_virtualenv_packs:  The two associated sys.prefixes are the network_essentials and dcfabric virtualenv paths

        NOTE: When multiple prefixes are specified for a single api daemon, the first listed prefix will the be virtualenv which the api daemon will start up under.  When an api daemon is 'restart'-ed, the config file is updated and the last restarted api daemon's prefix will be listed first.  The api deamon will be restarted during a pip install of a given virtualenv and which ever virtualenv is updated last we be the preferred virtualenv to service the rest of the configured virtualenvs.  

    - The 'ns_port = <tcp port #>' configuration is optional and no longer required.  If specified, then a pyswitchlib_ns_daemon will be luanched as well as the configured api daemons and pyswitchlib assets will use the name server daemon to lookup which api daemons to use.

    -  When the ns_port configuration is not specified, then a file is maintained to list which api daemons are running and how to connect to them.  Pyswitchlib assets will look up this file to connect to the proper api daemon.  The file is located at /tmp/.pswitchlib_ns_daemon.uri.

Caveats:
======
- Any python virtualenv that is not found in the config file will try to connect to the default API daemon that is started on the host's base python. 

Unit Test Installation:
================
- /opt/brocade/bwc-topology/bin/pip install -U git+https://badaniya@github.com/StackStorm/PySwitchLib@api_broker
- /opt/stackstorm/virtualenvs/network_essentials/bin/pip install -U git+https://badaniya@github.com/StackStorm/PySwitchLib@api_broker
- /opt/stackstorm/virtualenvs/dcfabric/bin/pip install -U git+https://badaniya@github.com/StackStorm/PySwitchLib@api_broker

Observations:
=========
- 760MB used by api broker (with all pyswitchlib apis and all current versions of pyangbindings loaded)
- Each pyswitchlib asset instance is now down to ~15-20MB.

Memory Profiler:
=============
BEFORE API DAEMON:
Line #    Mem usage    Increment   Line Contents
================================================
     3   10.320 MiB    0.000 MiB   @profile
     4                             def main():
     5  342.250 MiB  **331.930 MiB       import pyswitchlib.asset**
     6  342.250 MiB    0.000 MiB       import requests
     7  342.250 MiB    0.000 MiB       from pprint import pprint
     8  342.250 MiB    0.000 MiB       import sys
     9                                 
    10                                 #import time
    11  342.250 MiB    0.000 MiB       import logging
    12                                 
    13  342.250 MiB    0.000 MiB       rb = '69'
    14  342.250 MiB    0.000 MiB       nos1 = None
    15  342.250 MiB    0.000 MiB       slx1 = None
    16                                 
    17  342.250 MiB    0.000 MiB       logging.basicConfig(level=logging.INFO,                                                                                                                                     
    18  342.250 MiB    0.000 MiB                           format='%(levelname)-5s : %(asctime)s : %(name)s::%(threadName)-8s : %(message)s',                                                                      
    19                                                     )
    20                                 
    21  342.285 MiB    0.035 MiB       logging.info ("Starting test_pyswitchlib.py ...")
    22                                 
    23  342.285 MiB    0.000 MiB       logging.getLogger("requests").setLevel(logging.WARNING)                                                                                                                     
    24  342.285 MiB    0.000 MiB       logging.getLogger("dicttoxml").setLevel(logging.WARNING)
    25  342.285 MiB    0.000 MiB       try:
    26  342.898 MiB    0.613 MiB           nos1 = pyswitchlib.asset.Asset(ip_addr='10.24.81.125', auth=('admin', 'password'))
    27  342.953 MiB    0.055 MiB           slx1 = pyswitchlib.asset.Asset(ip_addr='10.24.84.125', auth=('admin', 'password'))
    28                                 
    29  342.953 MiB    0.000 MiB           logging.info('Asset: nos1 => ' + nos1.get_os_type() + ' ' + nos1.get_os_version() + ' ' + nos1.get_os_full_version())
    30  342.953 MiB    0.000 MiB           logging.info('Asset: slx1 => ' + slx1.get_os_type() + ' ' + slx1.get_os_version() + ' ' + slx1.get_os_full_version())
    31                                 
    32                                 except pyswitchlib.asset.RestInterfaceError:
    33                                     print('RestInterfaceError')
    34                                     raise RestInterfaceError
    35                                 except requests.exceptions.ConnectionError:
    36                                     print('ConnectionError')
    37                                     raise requests.exceptions.ConnectionError
    38                                 except ValueError:
    39                                     print('ValueError')
    40                                     raise RestInterfaceError
    41                                 
    42  444.613 MiB  **101.660 MiB       pprint(nos1.vlan_create(vlan='100'))**
    43  447.961 MiB    3.348 MiB       pprint(nos1.interface_ve_create(ve='100'))
    44  449.953 MiB    1.992 MiB       pprint(nos1.interface_ve_update(ve='100', global_ve_shutdown=False))
    45  450.168 MiB    0.215 MiB       pprint(nos1.interface_ve_update(ve='100', global_ve_shutdown=True))
    46  450.168 MiB    0.000 MiB       pprint(nos1.interface_ve_delete(ve='100'))
    47  454.230 MiB    4.062 MiB       pprint(nos1.vlan_delete(vlan='100'))
    48                                 
    49  549.762 MiB   **95.531 MiB       pprint(slx1.vlan_create(vlan='100'))**
    50  564.441 MiB   14.680 MiB       pprint(slx1.interface_ve_create(ve='100'))
    51  572.512 MiB    8.070 MiB       pprint(slx1.interface_ve_update(ve='100', global_ve_shutdown=False))
    52  572.512 MiB    0.000 MiB       pprint(slx1.interface_ve_update(ve='100', global_ve_shutdown=True))
    53  572.012 MiB   -0.500 MiB       pprint(slx1.interface_ve_delete(ve='100'))
    54  572.012 MiB    0.000 MiB       pprint(slx1.vlan_delete(vlan='100'))

AFTER API DAEMON:
Line #    Mem usage    Increment   Line Contents
================================================
     3   10.316 MiB    0.000 MiB   @profile
     4                             def main():
     5   18.641 MiB    **8.324 MiB       import pyswitchlib.asset**
     6   18.641 MiB    0.000 MiB       import requests
     7   18.641 MiB    0.000 MiB       from pprint import pprint
     8   18.641 MiB    0.000 MiB       import sys
     9                                 
    10                                 #import time
    11   18.641 MiB    0.000 MiB       import logging
    12                                 
    13   18.641 MiB    0.000 MiB       rb = '69'
    14   18.641 MiB    0.000 MiB       nos1 = None
    15   18.641 MiB    0.000 MiB       slx1 = None
    16                                 
    17   18.641 MiB    0.000 MiB       logging.basicConfig(level=logging.INFO,                                                                                                                                     
    18   18.641 MiB    0.000 MiB                           format='%(levelname)-5s : %(asctime)s : %(name)s::%(threadName)-8s : %(message)s',                                                                      
    19                                                     )
    20                                 
    21   18.668 MiB    0.027 MiB       logging.info ("Starting test_pyswitchlib.py ...")
    22                                 
    23   18.668 MiB    0.000 MiB       logging.getLogger("requests").setLevel(logging.WARNING)                                                                                                                     
    24   18.668 MiB    0.000 MiB       logging.getLogger("dicttoxml").setLevel(logging.WARNING)
    25   18.668 MiB    0.000 MiB       try:
    26   36.688 MiB   **18.020 MiB           nos1 = pyswitchlib.asset.Asset(ip_addr='10.24.81.125', auth=('admin', 'password'))**
    27   51.844 MiB   **15.156 MiB           slx1 = pyswitchlib.asset.Asset(ip_addr='10.24.84.125', auth=('admin', 'password'))**
    28                                 
    29   51.844 MiB    0.000 MiB           logging.info('Asset: nos1 => ' + nos1.get_os_type() + ' ' + nos1.get_os_version() + ' ' + nos1.get_os_full_version())
    30   51.844 MiB    0.000 MiB           logging.info('Asset: slx1 => ' + slx1.get_os_type() + ' ' + slx1.get_os_version() + ' ' + slx1.get_os_full_version())
    31                                 
    32                                 except pyswitchlib.asset.RestInterfaceError:
    33                                     print('RestInterfaceError')
    34                                     raise RestInterfaceError
    35                                 except requests.exceptions.ConnectionError:
    36                                     print('ConnectionError')
    37                                     raise requests.exceptions.ConnectionError
    38                                 except ValueError:
    39                                     print('ValueError')
    40                                     raise RestInterfaceError
    41                                 
    42   60.602 MiB    **8.758 MiB       pprint(nos1.vlan_create(vlan='100'))**
    43   60.605 MiB    0.004 MiB       pprint(nos1.interface_ve_create(ve='100'))
    44   60.605 MiB    0.000 MiB       pprint(nos1.interface_ve_update(ve='100', global_ve_shutdown=False))
    45   60.605 MiB    0.000 MiB       pprint(nos1.interface_ve_update(ve='100', global_ve_shutdown=True))
    46   60.605 MiB    0.000 MiB       pprint(nos1.interface_ve_delete(ve='100'))
    47   60.605 MiB    0.000 MiB       pprint(nos1.vlan_delete(vlan='100'))
    48                                 
    49   66.492 MiB    **5.887 MiB       pprint(slx1.vlan_create(vlan='100'))**
    50   66.492 MiB    0.000 MiB       pprint(slx1.interface_ve_create(ve='100'))
    51   66.492 MiB    0.000 MiB       pprint(slx1.interface_ve_update(ve='100', global_ve_shutdown=False))
    52   66.492 MiB    0.000 MiB       pprint(slx1.interface_ve_update(ve='100', global_ve_shutdown=True))
    53   66.492 MiB    0.000 MiB       pprint(slx1.interface_ve_delete(ve='100'))
    54   66.492 MiB    0.000 MiB       pprint(slx1.vlan_delete(vlan='100'))

API Broker:
=======
top - 08:57:34 up 74 days,  1:17,  0 users,  load average: 0.93, 1.29, 1.31
Tasks:   2 total,   0 running,   2 sleeping,   0 stopped,   0 zombie
%Cpu(s): 30.9 us,  2.3 sy,  0.1 ni, 66.7 id,  0.0 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem :  8160508 total,   182436 free,  3824400 used,  4153672 buff/cache
KiB Swap:  8377340 total,  7858120 free,   519220 used.  3867356 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                        
 3457 badaniya  20   0   84300   4736   1860 S   0.0  0.1   0:05.35 python                                                                                                         
 8448 badaniya  20   0 1559680 759164   9760 S   0.0  9.3   9:25.46 python

### CODE REVIEW

File: pyswitchlib.conf.default [hunk 1]
```diff
@@ -0,0 +1 @@
+ns_port = 9090
```
Comment (samirsss) [inline pyswitchlib.conf.default]: more of a question - What's this port for? Do we need to use a port for this purpose?

Comment (samirsss) [reply pyswitchlib.conf.default]: We should also document this, to make sure that the api-daemon startup doesnt fail if the port is occupied

Comment (badaniya) [reply pyswitchlib.conf.default]: The port is the name sever port for the daemon that will allow assets to connect it.  Should documentation be in the README and/or do we need something more?

Comment (samirsss) [reply pyswitchlib.conf.default]: We will have to check how the installation/configuration works when we have not just multiple packs, but services like bwc-topology service use pyswitchlib. Its going to be an issue when we have libraries will be requiring interface config and cause conflicts and failures.

Lets discuss this on Monday. Will setup a meeting for this.

**Pull request merged**